File: dish_forge.go
// dish_forge.go - Генератор случайных блюд на Go (CLI)
package main

import (
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

type Dish struct {
	Name         string   `json:"name"`
	Type         string   `json:"type"`
	Cuisine      string   `json:"cuisine"`
	Time         int      `json:"time"`
	Difficulty   string   `json:"difficulty"`
	Description  string   `json:"description"`
	Ingredients  []string `json:"ingredients"`
	Instructions []string `json:"instructions"`
}

type Forge struct {
	Dishes    []Dish   `json:"dishes"`
	Favorites []string `json:"favorites"`
}

const dataFile = "dishes.json"
const favFile = "favorites.json"

func loadForge() *Forge {
	var f Forge
	file, err := os.ReadFile(dataFile)
	if err != nil {
		f.Dishes = defaultDishes()
		f.Favorites = []string{}
		saveForge(&f)
		return &f
	}
	err = json.Unmarshal(file, &f)
	if err != nil {
		f.Dishes = defaultDishes()
		f.Favorites = []string{}
	}
	// load favorites separately
	favData, err := os.ReadFile(favFile)
	if err == nil {
		var favs []string
		json.Unmarshal(favData, &favs)
		f.Favorites = favs
	}
	return &f
}

func saveForge(f *Forge) {
	data, _ := json.MarshalIndent(f.Dishes, "", "  ")
	os.WriteFile(dataFile, data, 0644)
	favData, _ := json.MarshalIndent(f.Favorites, "", "  ")
	os.WriteFile(favFile, favData, 0644)
}

func defaultDishes() []Dish {
	return []Dish{
		{"Овсянка с ягодами", "breakfast", "russian", 10, "easy",
			"Полезный завтрак из овсяных хлопьев с ягодами и мёдом.",
			[]string{"Овсяные хлопья 100г", "Молоко 200мл", "Ягоды 50г", "Мёд 1 ст.л."},
			[]string{"Вскипятить молоко", "Добавить хлопья, варить 5 мин", "Добавить ягоды и мёд"}},
		{"Паста Карбонара", "lunch", "italian", 25, "medium",
			"Классическая итальянская паста с беконом и пармезаном.",
			[]string{"Паста 200г", "Бекон 100г", "Яйца 2 шт", "Пармезан 50г", "Сливки 100мл"},
			[]string{"Отварить пасту", "Обжарить бекон", "Смешать яйца, сыр и сливки", "Соединить с пастой"}},
		{"Суши Филадельфия", "dinner", "japanese", 40, "hard",
			"Роллы с лососем, сливочным сыром и авокадо.",
			[]string{"Рис для суши 200г", "Нори 4 шт", "Лосось 150г", "Сливочный сыр 100г", "Авокадо 1 шт"},
			[]string{"Сварить рис", "Нарезать лосось и авокадо", "Собрать роллы", "Нарезать и подавать"}},
	}
}

func (f *Forge) generate(dishType, cuisine, difficulty string, maxTime int) *Dish {
	filtered := f.Dishes
	if dishType != "" {
		filtered = filterByType(filtered, dishType)
	}
	if cuisine != "" {
		filtered = filterByCuisine(filtered, cuisine)
	}
	if maxTime > 0 {
		filtered = filterByTime(filtered, maxTime)
	}
	if difficulty != "" {
		filtered = filterByDifficulty(filtered, difficulty)
	}
	if len(filtered) == 0 {
		return nil
	}
	rand.Seed(time.Now().UnixNano())
	return &filtered[rand.Intn(len(filtered))]
}

func filterByType(dishes []Dish, t string) []Dish {
	var res []Dish
	for _, d := range dishes {
		if d.Type == t {
			res = append(res, d)
		}
	}
	return res
}
func filterByCuisine(dishes []Dish, c string) []Dish {
	var res []Dish
	for _, d := range dishes {
		if d.Cuisine == c {
			res = append(res, d)
		}
	}
	return res
}
func filterByTime(dishes []Dish, maxTime int) []Dish {
	var res []Dish
	for _, d := range dishes {
		if d.Time <= maxTime {
			res = append(res, d)
		}
	}
	return res
}
func filterByDifficulty(dishes []Dish, diff string) []Dish {
	var res []Dish
	for _, d := range dishes {
		if d.Difficulty == diff {
			res = append(res, d)
		}
	}
	return res
}

func (f *Forge) generateMenu(count int, dishType, cuisine, difficulty string, maxTime int) []Dish {
	available := f.Dishes
	if dishType != "" {
		available = filterByType(available, dishType)
	}
	if cuisine != "" {
		available = filterByCuisine(available, cuisine)
	}
	if maxTime > 0 {
		available = filterByTime(available, maxTime)
	}
	if difficulty != "" {
		available = filterByDifficulty(available, difficulty)
	}
	if len(available) < count {
		count = len(available)
	}
	rand.Seed(time.Now().UnixNano())
	shuffled := make([]Dish, len(available))
	copy(shuffled, available)
	rand.Shuffle(len(shuffled), func(i, j int) { shuffled[i], shuffled[j] = shuffled[j], shuffled[i] })
	return shuffled[:count]
}

func (f *Forge) addFavorite(name string) bool {
	for _, n := range f.Favorites {
		if n == name {
			return false
		}
	}
	f.Favorites = append(f.Favorites, name)
	saveForge(f)
	return true
}

func (f *Forge) removeFavorite(name string) bool {
	for i, n := range f.Favorites {
		if n == name {
			f.Favorites = append(f.Favorites[:i], f.Favorites[i+1:]...)
			saveForge(f)
			return true
		}
	}
	return false
}

func (f *Forge) getFavorites() []Dish {
	var result []Dish
	for _, d := range f.Dishes {
		for _, n := range f.Favorites {
			if d.Name == n {
				result = append(result, d)
				break
			}
		}
	}
	return result
}

func formatDish(d Dish) string {
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("🍽️ %s\n", d.Name))
	sb.WriteString(fmt.Sprintf("  Тип: %s\n", d.Type))
	sb.WriteString(fmt.Sprintf("  Кухня: %s\n", d.Cuisine))
	sb.WriteString(fmt.Sprintf("  Время: %d мин\n", d.Time))
	sb.WriteString(fmt.Sprintf("  Сложность: %s\n", d.Difficulty))
	sb.WriteString(fmt.Sprintf("  Описание: %s\n", d.Description))
	sb.WriteString("  Ингредиенты:\n")
	for _, ing := range d.Ingredients {
		sb.WriteString(fmt.Sprintf("    - %s\n", ing))
	}
	sb.WriteString("  Приготовление:\n")
	for i, step := range d.Instructions {
		sb.WriteString(fmt.Sprintf("    %d. %s\n", i+1, step))
	}
	return sb.String()
}

func main() {
	var (
		cmd        string
		dishType   string
		cuisine    string
		maxTime    int
		difficulty string
		count      int
		name       string
	)
	flag.StringVar(&cmd, "cmd", "", "Команда: generate, menu, list, favorites")
	flag.StringVar(&dishType, "type", "", "Тип блюда")
	flag.StringVar(&cuisine, "cuisine", "", "Кухня")
	flag.IntVar(&maxTime, "time", 0, "Макс. время")
	flag.StringVar(&difficulty, "difficulty", "", "Сложность")
	flag.IntVar(&count, "count", 3, "Количество блюд в меню")
	flag.StringVar(&name, "name", "", "Название блюда для избранного")
	flag.Parse()

	forge := loadForge()

	switch cmd {
	case "generate":
		dish := forge.generate(dishType, cuisine, difficulty, maxTime)
		if dish != nil {
			fmt.Println(formatDish(*dish))
		} else {
			fmt.Println("❌ Блюда не найдены")
		}
	case "menu":
		dishes := forge.generateMenu(count, dishType, cuisine, difficulty, maxTime)
		if len(dishes) > 0 {
			fmt.Printf("🍽️ Меню из %d блюд:\n", len(dishes))
			for i, d := range dishes {
				fmt.Printf("\n%d. %s (%s, %s, %d мин, %s)\n", i+1, d.Name, d.Type, d.Cuisine, d.Time, d.Difficulty)
				fmt.Printf("   %s\n", d.Description)
			}
		} else {
			fmt.Println("❌ Недостаточно блюд")
		}
	case "list":
		dishes := forge.Dishes
		if dishType != "" {
			dishes = filterByType(dishes, dishType)
		}
		if cuisine != "" {
			dishes = filterByCuisine(dishes, cuisine)
		}
		for _, d := range dishes {
			fmt.Printf("%s (%s, %s, %d мин)\n", d.Name, d.Type, d.Cuisine, d.Time)
		}
	case "favorites":
		if name != "" {
			if forge.addFavorite(name) {
				fmt.Printf("✅ '%s' добавлен в избранное\n", name)
			} else {
				fmt.Printf("⚠️ '%s' уже в избранном или не существует\n", name)
			}
		} else {
			favs := forge.getFavorites()
			if len(favs) > 0 {
				fmt.Println("⭐ Избранное:")
				for _, d := range favs {
					fmt.Printf("  %s (%s, %s)\n", d.Name, d.Type, d.Cuisine)
				}
			} else {
				fmt.Println("Избранное пусто")
			}
		}
	default:
		interactiveMode(forge)
	}
}

func interactiveMode(f *Forge) {
	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Println("\n🍳 DishForge - Генератор блюд (интерактивный)")
		fmt.Println("1. Сгенерировать случайное блюдо")
		fmt.Println("2. Сгенерировать меню")
		fmt.Println("3. Список всех блюд")
		fmt.Println("4. Избранное")
		fmt.Println("0. Выход")
		fmt.Print("Выберите действие: ")
		scanner.Scan()
		choice := scanner.Text()
		switch choice {
		case "0":
			return
		case "1":
			fmt.Print("Тип (Enter пропустить): ")
			scanner.Scan()
			t := scanner.Text()
			fmt.Print("Кухня (Enter пропустить): ")
			scanner.Scan()
			c := scanner.Text()
			fmt.Print("Макс. время (мин, Enter пропустить): ")
			scanner.Scan()
			timeStr := scanner.Text()
			mt := 0
			if timeStr != "" {
				mt, _ = strconv.Atoi(timeStr)
			}
			fmt.Print("Сложность (easy/medium/hard, Enter пропустить): ")
			scanner.Scan()
			diff := scanner.Text()
			dish := f.generate(t, c, diff, mt)
			if dish != nil {
				fmt.Println(formatDish(*dish))
			} else {
				fmt.Println("❌ Блюда не найдены")
			}
		case "2":
			fmt.Print("Количество блюд (по умолчанию 3): ")
			scanner.Scan()
			cntStr := scanner.Text()
			cnt := 3
			if cntStr != "" {
				cnt, _ = strconv.Atoi(cntStr)
			}
			fmt.Print("Тип (Enter пропустить): ")
			scanner.Scan()
			t := scanner.Text()
			fmt.Print("Кухня (Enter пропустить): ")
			scanner.Scan()
			c := scanner.Text()
			fmt.Print("Макс. время (мин, Enter пропустить): ")
			scanner.Scan()
			timeStr := scanner.Text()
			mt := 0
			if timeStr != "" {
				mt, _ = strconv.Atoi(timeStr)
			}
			fmt.Print("Сложность (Enter пропустить): ")
			scanner.Scan()
			diff := scanner.Text()
			dishes := f.generateMenu(cnt, t, c, diff, mt)
			if len(dishes) > 0 {
				fmt.Printf("🍽️ Меню из %d блюд:\n", len(dishes))
				for i, d := range dishes {
					fmt.Printf("\n%d. %s (%s, %s, %d мин, %s)\n", i+1, d.Name, d.Type, d.Cuisine, d.Time, d.Difficulty)
					fmt.Printf("   %s\n", d.Description)
				}
			} else {
				fmt.Println("❌ Недостаточно блюд")
			}
		case "3":
			fmt.Print("Тип (Enter пропустить): ")
			scanner.Scan()
			t := scanner.Text()
			fmt.Print("Кухня (Enter пропустить): ")
			scanner.Scan()
			c := scanner.Text()
			dishes := f.Dishes
			if t != "" {
				dishes = filterByType(dishes, t)
			}
			if c != "" {
				dishes = filterByCuisine(dishes, c)
			}
			for _, d := range dishes {
				fmt.Printf("  %s (%s, %s, %d мин)\n", d.Name, d.Type, d.Cuisine, d.Time)
			}
		case "4":
			favs := f.getFavorites()
			fmt.Println("⭐ Избранное:")
			if len(favs) > 0 {
				for _, d := range favs {
					fmt.Printf("  %s (%s, %s)\n", d.Name, d.Type, d.Cuisine)
				}
			} else {
				fmt.Println("  Пусто")
			}
			fmt.Println("\n1. Добавить в избранное")
			fmt.Println("2. Удалить из избранного")
			fmt.Println("3. Назад")
			fmt.Print("Выберите: ")
			scanner.Scan()
			sub := scanner.Text()
			if sub == "1" {
				fmt.Print("Название блюда: ")
				scanner.Scan()
				name := scanner.Text()
				if f.addFavorite(name) {
					fmt.Printf("✅ '%s' добавлен\n", name)
				} else {
					fmt.Printf("⚠️ '%s' уже в избранном или не существует\n", name)
				}
			} else if sub == "2" {
				fmt.Print("Название блюда: ")
				scanner.Scan()
				name := scanner.Text()
				if f.removeFavorite(name) {
					fmt.Printf("✅ '%s' удалён\n", name)
				} else {
					fmt.Printf("❌ '%s' не найден\n", name)
				}
			}
		default:
			fmt.Println("Неверный выбор")
		}
	}
}

File: dish_forge.rs
// dish_forge.rs - Генератор случайных блюд на Rust (CLI)
use serde::{Serialize, Deserialize};
use std::collections::HashSet;
use std::fs;
use std::io::{self, Write, BufRead};
use std::path::Path;
use std::str::FromStr;
use rand::seq::SliceRandom;
use rand::thread_rng;

#[derive(Serialize, Deserialize, Clone)]
struct Dish {
    name: String,
    dish_type: String,
    cuisine: String,
    time: u32,
    difficulty: String,
    description: String,
    ingredients: Vec<String>,
    instructions: Vec<String>,
}

#[derive(Serialize, Deserialize)]
struct Forge {
    dishes: Vec<Dish>,
    favorites: Vec<String>,
}

const DATA_FILE: &str = "dishes.json";
const FAV_FILE: &str = "favorites.json";

impl Forge {
    fn load() -> Self {
        let mut f = Forge {
            dishes: vec![],
            favorites: vec![],
        };
        if Path::new(DATA_FILE).exists() {
            if let Ok(data) = fs::read_to_string(DATA_FILE) {
                if let Ok(dishes) = serde_json::from_str(&data) {
                    f.dishes = dishes;
                }
            }
        }
        if f.dishes.is_empty() {
            f.dishes = default_dishes();
            f.save();
        }
        if Path::new(FAV_FILE).exists() {
            if let Ok(data) = fs::read_to_string(FAV_FILE) {
                if let Ok(favs) = serde_json::from_str(&data) {
                    f.favorites = favs;
                }
            }
        }
        f
    }

    fn save(&self) {
        let data = serde_json::to_string_pretty(&self.dishes).unwrap();
        fs::write(DATA_FILE, data).unwrap();
        let fav_data = serde_json::to_string_pretty(&self.favorites).unwrap();
        fs::write(FAV_FILE, fav_data).unwrap();
    }

    fn generate(&self, dish_type: Option<&str>, cuisine: Option<&str>, max_time: Option<u32>, difficulty: Option<&str>) -> Option<Dish> {
        let mut filtered = self.dishes.clone();
        if let Some(t) = dish_type {
            filtered.retain(|d| d.dish_type == t);
        }
        if let Some(c) = cuisine {
            filtered.retain(|d| d.cuisine == c);
        }
        if let Some(t) = max_time {
            filtered.retain(|d| d.time <= t);
        }
        if let Some(d) = difficulty {
            filtered.retain(|dsh| dsh.difficulty == d);
        }
        if filtered.is_empty() {
            return None;
        }
        let mut rng = thread_rng();
        filtered.choose(&mut rng).cloned()
    }

    fn generate_menu(&self, count: usize, dish_type: Option<&str>, cuisine: Option<&str>, max_time: Option<u32>, difficulty: Option<&str>) -> Vec<Dish> {
        let mut available = self.dishes.clone();
        if let Some(t) = dish_type {
            available.retain(|d| d.dish_type == t);
        }
        if let Some(c) = cuisine {
            available.retain(|d| d.cuisine == c);
        }
        if let Some(t) = max_time {
            available.retain(|d| d.time <= t);
        }
        if let Some(d) = difficulty {
            available.retain(|dsh| dsh.difficulty == d);
        }
        let count = count.min(available.len());
        let mut rng = thread_rng();
        available.shuffle(&mut rng);
        available[..count].to_vec()
    }

    fn add_favorite(&mut self, name: &str) -> bool {
        if self.favorites.contains(&name.to_string()) {
            return false;
        }
        self.favorites.push(name.to_string());
        self.save();
        true
    }

    fn remove_favorite(&mut self, name: &str) -> bool {
        let idx = self.favorites.iter().position(|n| n == name);
        if let Some(idx) = idx {
            self.favorites.remove(idx);
            self.save();
            return true;
        }
        false
    }

    fn get_favorites(&self) -> Vec<Dish> {
        let mut result = vec![];
        for d in &self.dishes {
            if self.favorites.contains(&d.name) {
                result.push(d.clone());
            }
        }
        result
    }
}

fn default_dishes() -> Vec<Dish> {
    vec![
        Dish {
            name: "Овсянка с ягодами".to_string(),
            dish_type: "breakfast".to_string(),
            cuisine: "russian".to_string(),
            time: 10,
            difficulty: "easy".to_string(),
            description: "Полезный завтрак из овсяных хлопьев с ягодами и мёдом.".to_string(),
            ingredients: vec!["Овсяные хлопья 100г".to_string(), "Молоко 200мл".to_string(), "Ягоды 50г".to_string(), "Мёд 1 ст.л.".to_string()],
            instructions: vec!["Вскипятить молоко".to_string(), "Добавить хлопья, варить 5 мин".to_string(), "Добавить ягоды и мёд".to_string()],
        },
        Dish {
            name: "Паста Карбонара".to_string(),
            dish_type: "lunch".to_string(),
            cuisine: "italian".to_string(),
            time: 25,
            difficulty: "medium".to_string(),
            description: "Классическая итальянская паста с беконом и пармезаном.".to_string(),
            ingredients: vec!["Паста 200г".to_string(), "Бекон 100г".to_string(), "Яйца 2 шт".to_string(), "Пармезан 50г".to_string(), "Сливки 100мл".to_string()],
            instructions: vec!["Отварить пасту".to_string(), "Обжарить бекон".to_string(), "Смешать яйца, сыр и сливки".to_string(), "Соединить с пастой".to_string()],
        },
    ]
}

fn format_dish(d: &Dish) -> String {
    let mut lines = vec![];
    lines.push(format!("🍽️ {}", d.name));
    lines.push(format!("  Тип: {}", d.dish_type));
    lines.push(format!("  Кухня: {}", d.cuisine));
    lines.push(format!("  Время: {} мин", d.time));
    lines.push(format!("  Сложность: {}", d.difficulty));
    lines.push(format!("  Описание: {}", d.description));
    lines.push("  Ингредиенты:".to_string());
    for ing in &d.ingredients {
        lines.push(format!("    - {}", ing));
    }
    lines.push("  Приготовление:".to_string());
    for (i, step) in d.instructions.iter().enumerate() {
        lines.push(format!("    {}. {}", i+1, step));
    }
    lines.join("\n")
}

fn read_line(prompt: &str) -> String {
    print!("{}", prompt);
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    input.trim().to_string()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        interactive_mode();
        return;
    }
    let mut forge = Forge::load();
    match args[1].as_str() {
        "generate" => {
            let mut dish_type = None;
            let mut cuisine = None;
            let mut max_time = None;
            let mut difficulty = None;
            let mut i = 2;
            while i < args.len() {
                match args[i].as_str() {
                    "--type" => { dish_type = Some(args[i+1].clone()); i += 2; }
                    "--cuisine" => { cuisine = Some(args[i+1].clone()); i += 2; }
                    "--time" => { max_time = Some(args[i+1].parse().unwrap_or(0)); i += 2; }
                    "--difficulty" => { difficulty = Some(args[i+1].clone()); i += 2; }
                    _ => { i += 1; }
                }
            }
            if let Some(dish) = forge.generate(dish_type.as_deref(), cuisine.as_deref(), max_time, difficulty.as_deref()) {
                println!("{}", format_dish(&dish));
            } else {
                println!("❌ Блюда не найдены");
            }
        }
        "menu" => {
            let mut count = 3;
            let mut dish_type = None;
            let mut cuisine = None;
            let mut max_time = None;
            let mut difficulty = None;
            let mut i = 2;
            while i < args.len() {
                match args[i].as_str() {
                    "--count" => { count = args[i+1].parse().unwrap_or(3); i += 2; }
                    "--type" => { dish_type = Some(args[i+1].clone()); i += 2; }
                    "--cuisine" => { cuisine = Some(args[i+1].clone()); i += 2; }
                    "--time" => { max_time = Some(args[i+1].parse().unwrap_or(0)); i += 2; }
                    "--difficulty" => { difficulty = Some(args[i+1].clone()); i += 2; }
                    _ => { i += 1; }
                }
            }
            let dishes = forge.generate_menu(count, dish_type.as_deref(), cuisine.as_deref(), max_time, difficulty.as_deref());
            if !dishes.is_empty() {
                println!("🍽️ Меню из {} блюд:", dishes.len());
                for (i, d) in dishes.iter().enumerate() {
                    println!("\n{}. {} ({}, {}, {} мин, {})", i+1, d.name, d.dish_type, d.cuisine, d.time, d.difficulty);
                    println!("   {}", d.description);
                }
            } else {
                println!("❌ Недостаточно блюд");
            }
        }
        "list" => {
            let mut dish_type = None;
            let mut cuisine = None;
            let mut i = 2;
            while i < args.len() {
                match args[i].as_str() {
                    "--type" => { dish_type = Some(args[i+1].clone()); i += 2; }
                    "--cuisine" => { cuisine = Some(args[i+1].clone()); i += 2; }
                    _ => { i += 1; }
                }
            }
            let mut dishes = forge.dishes.clone();
            if let Some(t) = dish_type {
                dishes.retain(|d| d.dish_type == t);
            }
            if let Some(c) = cuisine {
                dishes.retain(|d| d.cuisine == c);
            }
            for d in dishes {
                println!("{} ({}, {}, {} мин)", d.name, d.dish_type, d.cuisine, d.time);
            }
        }
        "favorites" => {
            if args.len() > 3 && args[2] == "add" {
                let name = args[3].clone();
                if forge.add_favorite(&name) {
                    println!("✅ '{}' добавлен в избранное", name);
                } else {
                    println!("⚠️ '{}' уже в избранном или не существует", name);
                }
            } else if args.len() > 3 && args[2] == "remove" {
                let name = args[3].clone();
                if forge.remove_favorite(&name) {
                    println!("✅ '{}' удалён из избранного", name);
                } else {
                    println!("❌ '{}' не найден в избранном", name);
                }
            } else {
                let favs = forge.get_favorites();
                if !favs.is_empty() {
                    println!("⭐ Избранное:");
                    for d in favs {
                        println!("  {} ({}, {})", d.name, d.dish_type, d.cuisine);
                    }
                } else {
                    println!("Избранное пусто");
                }
            }
        }
        _ => interactive_mode(),
    }
}

fn interactive_mode() {
    let mut forge = Forge::load();
    let stdin = io::stdin();
    let mut stdout = io::stdout();
    loop {
        println!("\n🍳 DishForge - Генератор блюд (интерактивный)");
        println!("1. Сгенерировать случайное блюдо");
        println!("2. Сгенерировать меню");
        println!("3. Список всех блюд");
        println!("4. Избранное");
        println!("0. Выход");
        print!("Выберите действие: ");
        stdout.flush().unwrap();
        let mut choice = String::new();
        stdin.read_line(&mut choice).unwrap();
        match choice.trim() {
            "0" => break,
            "1" => {
                let t = read_line("Тип (Enter пропустить): ");
                let t = if t.is_empty() { None } else { Some(t.as_str()) };
                let c = read_line("Кухня (Enter пропустить): ");
                let c = if c.is_empty() { None } else { Some(c.as_str()) };
                let time_str = read_line("Макс. время (мин, Enter пропустить): ");
                let max_time = if time_str.is_empty() { None } else { Some(time_str.parse::<u32>().unwrap_or(0)) };
                let diff = read_line("Сложность (easy/medium/hard, Enter пропустить): ");
                let diff = if diff.is_empty() { None } else { Some(diff.as_str()) };
                if let Some(dish) = forge.generate(t, c, max_time, diff) {
                    println!("{}", format_dish(&dish));
                } else {
                    println!("❌ Блюда не найдены");
                }
            }
            "2" => {
                let count_str = read_line("Количество блюд (по умолчанию 3): ");
                let count = if count_str.is_empty() { 3 } else { count_str.parse::<usize>().unwrap_or(3) };
                let t = read_line("Тип (Enter пропустить): ");
                let t = if t.is_empty() { None } else { Some(t.as_str()) };
                let c = read_line("Кухня (Enter пропустить): ");
                let c = if c.is_empty() { None } else { Some(c.as_str()) };
                let time_str = read_line("Макс. время (мин, Enter пропустить): ");
                let max_time = if time_str.is_empty() { None } else { Some(time_str.parse::<u32>().unwrap_or(0)) };
                let diff = read_line("Сложность (Enter пропустить): ");
                let diff = if diff.is_empty() { None } else { Some(diff.as_str()) };
                let dishes = forge.generate_menu(count, t, c, max_time, diff);
                if !dishes.is_empty() {
                    println!("🍽️ Меню из {} блюд:", dishes.len());
                    for (i, d) in dishes.iter().enumerate() {
                        println!("\n{}. {} ({}, {}, {} мин, {})", i+1, d.name, d.dish_type, d.cuisine, d.time, d.difficulty);
                        println!("   {}", d.description);
                    }
                } else {
                    println!("❌ Недостаточно блюд");
                }
            }
            "3" => {
                let t = read_line("Тип (Enter пропустить): ");
                let c = read_line("Кухня (Enter пропустить): ");
                let mut dishes = forge.dishes.clone();
                if !t.is_empty() {
                    dishes.retain(|d| d.dish_type == t);
                }
                if !c.is_empty() {
                    dishes.retain(|d| d.cuisine == c);
                }
                for d in dishes {
                    println!("  {} ({}, {}, {} мин)", d.name, d.dish_type, d.cuisine, d.time);
                }
            }
            "4" => {
                let favs = forge.get_favorites();
                println!("⭐ Избранное:");
                if !favs.is_empty() {
                    for d in favs {
                        println!("  {} ({}, {})", d.name, d.dish_type, d.cuisine);
                    }
                } else {
                    println!("  Пусто");
                }
                println!("\n1. Добавить в избранное");
                println!("2. Удалить из избранного");
                println!("3. Назад");
                let sub = read_line("Выберите: ");
                if sub == "1" {
                    let name = read_line("Название блюда: ");
                    if forge.add_favorite(&name) {
                        println!("✅ '{}' добавлен", name);
                    } else {
                        println!("⚠️ '{}' уже в избранном или не существует", name);
                    }
                } else if sub == "2" {
                    let name = read_line("Название блюда: ");
                    if forge.remove_favorite(&name) {
                        println!("✅ '{}' удалён", name);
                    } else {
                        println!("❌ '{}' не найден", name);
                    }
                }
            }
            _ => println!("Неверный выбор"),
        }
    }
}
